### DISCUSSION TITLE
How does WebView2 inject JS into new pages in Window.open

### DISCUSSION THREAD

**Discussion by smartwei** [MicrosoftEdge/WebView2Feedback#3963] [Q&A]
My Winform application uses WebView2,
When I open page1.html in WebView2, the page1.html code looks like this:
```
<html>
<head>
<title>page1</title>
</head>
<body>
this is page 111
<script language=javascript>
var toUrl="http://localhost/KintechToGica22/page2.html";
window.open(toUrl,'openTest2','');
</script>
</body>
</html>
```
I use window.open in pag1.html to open a page page2.html,,the page2.html code looks like this:
```
<html>
<head>
<title>page2</title>
</head>
<body>
this is Page 222
</body>
</html>
```

But how do I control page2.html,
I want to do something with page2.html injection JS

**Comment by vmeganathan81** [comment]
It does not use window.open, please use Add host object

**Comment by vmeganathan81** [comment]
Its working for us,

Add Host Object, sample below

//Browse Host initialize during webview2 initialize

                BrowserHostObject browserHostObject = new BrowserHostObject();
                browserHostObject.MessageEvent += BrowserHostObject_MessageEvent; // Call back event
                webView2.CoreWebView2.AddHostObjectToScript("external", browserHostObject);
//Call back from web to .net application

        private void BrowserHostObject_MessageEvent(object sender, string e)
        {
                RaiseMessageEvent(e);            
        }
In HTML

        <button onclick="JsTest()">Callback_To_NET _App</button>
        <script>
                const external = chrome.webview.hostObjects.external;
                function JsTest() {
                        external.CallbackVB('This is a test');
                }
        </script>
Hope this helps

**Comment by smartwei** [comment]
@[vmeganathan81](https://github.com/vmeganathan81) 
Thank you for your reply

This page1. html is not my own page and I have no control over the content of this page.

As soon as I open page1. html, window.open is automatically executed and page2.html is opened. 
I just want to inject a JS into page2.html to do something

**Comment by david-risney** [comment]
You need to handle the [CoreWebView2.NewWindowRequested event](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2.newwindowrequested?view=webview2-dotnet-1.0.1210.39). By default without handling that event, new windows are opened in mini-browser-like windows. The end user can interact with these windows but the end developer cannot. If you would like to interact with that should handle the event.

You have several options which I have examples of in [new window sample code](https://github.com/david-risney/WebView2Samples/blob/9fb1cbc1b498e13764a3cd8bae5114f7876d3b06/SampleApps/webview2_sample_uwp/MainPage.xaml.cs#L51)

```c#
            switch (NewWindowBehaviorComboBox.SelectedIndex)
            {
                // Open mini browser window
                case 0:
                    // Default behavior. No change to args required.
                    break;

                // Send to default browser 
                case 1:
                    args.Handled = true;
                    // No need to wait for the launcher to finish sending the URI to the browser
                    // before we allow the WebView2 in our app to continue.
                    _ = Windows.System.Launcher.LaunchUriAsync(new Uri(args.Uri));
                    break;

                // Navigate in same WebView2
                case 2:
                    args.Handled = true;
                    args.NewWindow = sender;
                    break;

                // Open in different WebView2
                case 3:
                    args.Handled = true;
                    using (args.GetDeferral())
                    {
                        await WebView2Right.EnsureCoreWebView2Async();
                        args.NewWindow = WebView2Right.CoreWebView2;
                    }
                    break;

                // Ignore
                case 4:
                    args.Handled = true;
                    break;
            }
```

If you want handle the new window request in a different WebView2 control, you'll need to write app code to create a new WebView2. The sample above just uses an already created WebView2Right member. But for another application you might want to open a new app window and create a WebView2 in that window.

**Comment by smartwei** [comment]
[david-risney](https://github.com/david-risney)
Where variable "NewWindowBehaviorComboBox" and "WebView2Right" is defined
I'm using a Windows Form program,
I didn't find these two variables defined in the Demo

**Comment by david-risney** [comment]
The sample is a XAML app and the WebView2Right and the combo box are [defined in the XAML](https://github.com/david-risney/WebView2Samples/blob/9fb1cbc1b498e13764a3cd8bae5114f7876d3b06/SampleApps/webview2_sample_uwp/MainPage.xaml). For WinForms it will be very similar: you'll use a different WebView2 or create a new WebView2 control.
